students/daniel_carrasco/session05/mailroom.py

```diff
@@ -0,0 +1,67 @@
+#!/usr/bin/env python3
+from datetime import date
+import os
+import sys
+
+def main():
+    while True:
+        VALID_RESPONSES = (1,2,3,4)
+        try:
+            user_choice = int(input('\nChoose an action(1 - 4):\n\n\
+                    1 - Send a Thank You to a single donor.\n\
+                    2 - Create a Report.\n\
+                    3 - Send letters to all donors.\n\
+                    4 - Quit\n'))
+            arg_dict = {1:thankyou, 2:report, 3: letter,4:quit}
+            if arg_dict[user_choice] == 'quit':
+                sys.exit()
+            else:
+                arg_dict[user_choice]()
+        except ValueError:
+            print("Input must be an integer, try again.")
+        except KeyError:
+            print('Choice must be a menu input (1-4)')
+        continue
+
+def thankyou():
+    while True:
+        try:
+            choice = int(input('\nChoose an action(1-3):\n\
+                    1 - See Donor List.\n\
+                    2 - Enter Name.\n\
+                    3 - Quit submenu\n'))
+            if choice == 1:
+                for keys in donors.keys():
+                    print(keys.title())
+            if choice == 2:
+                new_name = input('Enter full name\n').title()
+                donors.setdefault(new_name,[0,1])
+            if choice == 3:
+                main()
+        except ValueError:
+            print("Input must be an integer, try again.")
+            continue
+
+def report():
+    row_format ="{:>15}" * 4
+    print(row_format.format('Name','Donation ($)','Amount','Average ($)'))
+    for key,value in donors.items():
```

Comment:
Use this one to implement comprehensions, it should be a relatively easy fit for the example provided as part of the exercise:

```
[print(donor) for donor in donors]
```

Reply:
One more comment on this one: If you run Pylint, it will complain that the _print()_ statement doesn't have a return, so your list comprehension will not really return a list. It will still work, but Pylint will make you feel bad about it :-)

One alternative is to take out the _print()_ statement a capture the output into a list for later printing. For example:

```
user_data = {
    "user01" : {
        "name" : "Robert",
        "last_name" : "Carnavaro"
    },
    "user02" : {
        "name" : "Adam",
        "last_name" : "Gouda"
    },
    "user03" : {
        "name" : "Luisa",
        "last_name" : "Ramirez"
    },
}

user_strings = [f'For user {user_name} name is {value["name"]} and last name {value["last_name"]}' for user_name, value in user_data.items()]

for user_string in user_strings:
    print(user_string)
```